Fix post-oxidization change in `ConsolidateBlocks` behavior

ElePT opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
Fixes the behavioral difference introduced by #13368 that caused #13438. 

The original implementation of the pass used to set the `ConsolidateBlocks` internal decomposer calling `unitary_synthesis._decomposer_2q_from_basis_gates`, which, when `kak_basis` wasn't found, would return `None`. During the oxidization efforts, `unitary_synthesis._decomposer_2q_from_basis_gates` was replaced with inline code that didn't contemplate this path (always set a default decomposer), which in the reproducing example led to a longer output circuit (the gates were collected into a unitary block and then decomposed differently).

### Details and comments
Technically not a bugfix because the change wasn't released yet.

Using the example shown in the issue ->

`ConsolidateBlocks` output dag with 1.2.4:
![image](https://github.com/user-attachments/assets/d3d297fd-fc76-41f6-9ece-68d252596ad6)

`ConsolidateBlocks` output dag with 1.3.0rc1:
![image](https://github.com/user-attachments/assets/b56ab05e-e9ae-4f16-a730-944ebd491938)

`ConsolidateBlocks` output dag after this fix:
![image](https://github.com/user-attachments/assets/d3d297fd-fc76-41f6-9ece-68d252596ad6)

ElePT committed:
Set decomposer to None if kak_gates not found during ConsolidateBlocks initialization.

ElePT committed:
Add test that reproduces issue

mtreinish reviewed (dismissed):
Ah, good catch.it was a change in behavior on my consolidate blocks pr not unitary synthesis. I remember changing that logic at the time and thinking it wouldn't impact anything. I was coming at it from the mindset that consolidate blocks is too likely to skip consolidation when it shouldn't.

Realistically we should be relying on the xx decomposer with the rzz in the basis gates list, but there isn't an interface in rust yet to tell us the estimated number of 2q basis gates needed for a unitary, which is why I changed the logic here. Also that is not really a good meteic anyway, just more reason for me to keep working on #13419 and introduce a better approach.

ElePT committed:
Re-add rzz to list of kak gates

ElePT commented:
Changing the "default" to `None` now left the paths for `rzz` and `rzx` gates unaccounted for. That broke the unit test that was fixed in [e3124a6](https://github.com/Qiskit/qiskit/pull/13450/commits/e3124a615f7e89d1ae7589e48e0c860780c9272d). So, questions:
1. any specific reason we had not to have `rzz` in the list? (because I just re-added it)
2. assuming that we don't want to use the `XXDecomposer` for `rzx`. What do we want to have then, `None` or `TwoQubitBasisDecomposer(CXGate())`? It currently doesn't break any test but just to confirm.

ShellyGarion commented on qiskit/transpiler/passes/optimization/consolidate_blocks.py:
```diff
@@ -27,6 +29,7 @@
     "cz": CZGate(),
     "iswap": iSwapGate(),
     "ecr": ECRGate(),
+    "rxx": RXXGate(pi / 2),
```
why is the angle here fixed to pi/2 ?

mtreinish replied:
It's because that's what `UnitarySynthesis` uses if it's a `basis_gates` only setup: https://github.com/Qiskit/qiskit/blob/main/qiskit/transpiler/passes/synthesis/unitary_synthesis.py#L98 the idea is that it's trying to match the decomposer used for estimating to what the synthesis pass will use later.

The whole logic between the passes is more than a bit hacky. `ConsolidateBlocks` tries to guess how many 2q gates will be used to decide whether to consolidate the block into a unitary or not. This is strongly coupled to the behavior of `UnitarySynthesis` because that's what actually does the synthesis later. But none of this actually takes into account multiple possible decomposers/synthesis methods being supported by a target, or the presence of a non-default plugin being used.

mtreinish replied:
FWIW, I did some testing locally with this and in some cases using `TwoQubitBasisDecomposer(RXXGate(pi/2))` under estimates what the `XXDecomposer` estimates, 3 vs 4 gates. This might result in some blocks not getting consolidated which were before, but in practice I find it probably pretty unlikely since it only counts rzz gates when evaluating if the block's gates are more than the expected synthesis output.

ElePT commented:
I added a small refactoring on top of the previous changes to make the new implementation hopefully closer to the old one. Now:
- ~The default decomposer is `None` when no basis_gates/kak_gates are found (same as OG). This includes the `rzz` gate from the issue~ [EDIT: didn't work. Reverted that change]
- If an `rxx` gate is found, the decomposer will be  the `TwoQubitBasisDecomposer` with a `RXX(pi/2)` (same as OG)
- If an `rzx` gate is found, the decomposer will be the `TwoQubitBasisDecomposer` with a `CXGate` (same as backup decomposer from OG implementation)

I also added a comment that mentions the previous use of the XXDecomposer for the `rzx` case for clarity.

ElePT commented on qiskit/transpiler/passes/optimization/consolidate_blocks.py:
```diff
@@ -70,19 +81,22 @@ def __init__(
+            if kak_gate:
+                if isinstance(kak_gate, RZXGate):
+                    # Backup optimizer from original implementation.
+                    # Once XXDecomposer is migrated to rust, this line could be rewritten ot use it.
```
```suggestion
                    # Once XXDecomposer is migrated to rust, this line could be rewritten to use it.
```

ElePT committed:
Add path for rzx gate.

mtreinish reviewed (approved):
I'm fine with this it returns the previous state of logic. There might be some subtle differences in what gets consolidated or not with rzx based on differences between `XXDecomposer.num_basis_gates` and `TwoQubitBasisDecomposer`'s logic, but it should be close enough to the 1.2.x state of the logic that I'm not worried. 

I think this whole exercise reinforces for me how we need to improve the logic around block consolidation. #13419 will do this for the optimization loop, but we will need to think what makes the most sense for when we run `ConsolidateBlocks` prior to layout in the init stage.
